SMLHelper/Patchers/PrefabDatabasePatcher.cs

```diff
@@ -1,82 +1,134 @@
+                return true;
 
-            return true;
+            __result = prefab.GetGameObjectInternal();
+            return false;
         }
 
-        internal static bool TryGetPrefabFilename_Prefix(string classId, ref string filename, ref bool __result)
+        private static IPrefabRequest GetModPrefabAsync(string classId)
         {
-            if (ModPrefab.TryGetFromClassId(classId, out ModPrefab prefab))
+            if (!ModPrefab.TryGetFromClassId(classId, out ModPrefab prefab))
+                return null;
+
+            try
             {
-                filename = prefab.PrefabFileName;
-                __result = true;
-                return false;
+                // trying sync method for backward compatibility
+                if (prefab.GetGameObjectInternal() is GameObject go) // SUBNAUTICA_EXP TODO: remove for SN after async update
+                    return new LoadedPrefabRequest(go);
             }
-            return true;
+            catch (Exception) {}
+
+            return new ModPrefabRequest(prefab);
         }
 
+        [PatchUtils.Prefix]
+        [HarmonyPatch(typeof(PrefabDatabase), nameof(PrefabDatabase.GetPrefabAsync))]
         internal static bool GetPrefabAsync_Prefix(ref IPrefabRequest __result, string classId)
         {
-            if (ModPrefab.TryGetFromClassId(classId, out ModPrefab prefab))
-            { 
-                GameObject go = prefab.GetGameObjectInternal();
-                __result = new LoadedPrefabRequest(go);
+            __result ??= GetModPrefabAsync(classId);
+            return __result == null;
+        }
+
+
+        // transpiler for ProtobufSerializer.DeserializeObjectsAsync
+        private static IEnumerable<CodeInstruction> DeserializeObjectsAsync_Transpiler(IEnumerable<CodeInstruction> cins)
+        {
+            var originalMethod = AccessTools.Method(typeof(ProtobufSerializer), nameof(ProtobufSerializer.InstantiatePrefabAsync));
+
+            return new CodeMatcher(cins).
+                MatchForward(false, new CodeMatch(OpCodes.Call, originalMethod)).
+                SetOperandAndAdvance(AccessTools.Method(typeof(PrefabDatabasePatcher), nameof(_InstantiatePrefabAsync))).
+                InstructionEnumeration();
+        }
+
+        // calling this instead of InstantiatePrefabAsync in ProtobufSerializer.DeserializeObjectsAsync
+        private static IEnumerator _InstantiatePrefabAsync(ProtobufSerializer.GameObjectData gameObjectData, IOut<UniqueIdentifier> result)
+        {
+            if (GetModPrefabAsync(gameObjectData.ClassId) is IPrefabRequest request)
+            {
+                yield return request;
 
-                return false;
+                if (request.TryGetPrefab(out GameObject prefab))
+                {
+                    result.Set(UnityEngine.Object.Instantiate(prefab).GetComponent<UniqueIdentifier>());
+                    yield break;
+                }
```

Comment:
What happens if this `TryGetPrefab` returns false?

Reply:
Then `ProtobufSerializer.InstantiatePrefabAsync` in line 106 supposed to be used.